packages/yarnpkg-core/sources/Project.ts

```diff
@@ -425,16 +425,30 @@ export class Project {
   forgetResolution(descriptor: Descriptor): void;
   forgetResolution(locator: Locator): void;
   forgetResolution(dataStructure: Descriptor | Locator): void {
-    for (const [descriptorHash, locatorHash] of this.storedResolutions) {
-      const doDescriptorHashesMatch = `descriptorHash` in dataStructure
-        && dataStructure.descriptorHash === descriptorHash;
-      const doLocatorHashesMatch = `locatorHash` in dataStructure
-        && dataStructure.locatorHash === locatorHash;
-
-      if (doDescriptorHashesMatch || doLocatorHashesMatch) {
-        this.storedDescriptors.delete(descriptorHash);
-        this.storedResolutions.delete(descriptorHash);
-        this.originalPackages.delete(locatorHash);
+    // Forgetting a descriptor requires:
+    // - deleting the resolution of the descriptor
+    // - deleting the descriptor
+    if (`descriptorHash` in dataStructure) {
+      this.storedResolutions.delete(dataStructure.descriptorHash);
+      this.storedDescriptors.delete(dataStructure.descriptorHash);
```

Comment:
Shouldn't we still delete the locator if there's no other reference to it? That might be very expensive though, unless we build a reference map beforehand 🤔 

Basically, I'm worried that the remaining `originalPackage` will prevent the package metadata from being updated when calling `yarn up` (since the old ones would linger).

Reply:
Actually the perf is moot - `forgetResolution` is only called with a descriptor inside `yarn up`, outside of a critical loop. So it won't be a problem even if we iterate to find all the usage.

Reply:
Done. 👍